iOS-sample-app/Showcases/Screen Views/Paged Scroll Views/PagedScrollViewController.swift

```diff
@@ -0,0 +1,65 @@
+
+import UIKit
+import ContentSquare
+
+// This class showcases how to track screen views for each page of a paged scroll view. There are many ways to do such
+// views so your implementation might differ. The key aspects should be the same:
+// - Trigger a screen view when the scroll view is displayed
+// - Trigger a screen view every time the page changes
+class PagedScrollViewController: UIViewController, UIScrollViewDelegate {
+    private static func makePage(number: Int) -> UIView {
+        let page = Bundle.main.loadNibNamed("Page", owner: self, options: nil)?.first as! Page
+        page.pageNumber = number
+        return page
+    }
+
+    @IBOutlet weak var scrollView: UIScrollView!
+
+    private let page1 = makePage(number: 1)
+    private let page2 = makePage(number: 2)
+    private let page3 = makePage(number: 3)
+
+    private var currentPage: Int {
+        let pageWidth = scrollView.frame.size.width
+        let offset = scrollView.contentOffset.x
+        return Int(round(offset / pageWidth) + 1)
+    }
+    private var previousPage: Int = 1
+
+    override func viewDidLoad() {
+        scrollView.addSubview(page1)
+        scrollView.addSubview(page2)
+        scrollView.addSubview(page3)
+
+        let views: [String: UIView] = ["scrollview": scrollView, "page1": page1, "page2": page2, "page3": page3]
+        let verticalConstraints = NSLayoutConstraint.constraints(
+            withVisualFormat: "V:|[page1(==scrollview)]|",
+            options: [],
+            metrics: nil,
+            views: views)
+        let horizontalConstraints = NSLayoutConstraint.constraints(
+            withVisualFormat: "H:|[page1(==scrollview)][page2(==scrollview)][page3(==scrollview)]|",
+            options: [.alignAllTop, .alignAllBottom],
+            metrics: nil,
+            views: views)
+        NSLayoutConstraint.activate(verticalConstraints + horizontalConstraints)
+    }
```

Comment:
I'm not quiet sure how you want me to improve this. I've taken a look to HorizontalScrollViewPagerController and it doesn't seem any cleaner to me.

The differences are basically these:
- I use a xib, you create views in code
- I use constraints, you set frames in code

The HorizontalScrollViewPagerController.swift file takes 30 more lines of code than I do at creating views, so the approach seems to get more in the way of teaching how to use the SDK IMO.

I believe I should keep my xib, as it gives a quick and visual idea of how the view looks, which is IMO better than having to read code about labels being added. Basically, an image tells a thousand words 😛. Also, I need to keep my Page.swift file which you told me to remove. Indeed it is the backing class of that view, and has an outlet to a label we want to make variable. It also is a super small file so I don't think it makes anything hard to understand. The main reason why HorizontalScrollViewPagerController is too long is because it creates views in code, so I really think this xib is a plus here.

So the thing that makes my code harder to read is the constraints. I understand that the Visual Formatting Language isn't always the best approach, but in this particular case it seems pretty intuitive and concise to me. The problem with your approach is that you hardcode frames, so as soon as the device is rotated the view completely breaks. I believe this app should be well made, so I don't want to go with that.

So here are the options regarding layout:
- Keep it as is -> not great since VFL can be mysterious if you don't know about it
- Use anchors to do auto layout instead -> would be easier to understand but take more of lines of code, not great.
- Make frames in code and have them break on rotation -> no go for me. Would also break when resizing if we ever let this app work on iPad.
- Make frames in code and try to make them not break thanks to layoutSubviews -> I'm going to try this approach

What do you think of this ?

Reply:
I am not sure that, for a thing as simple as a blank page with two text fields, we need to have 2 additional files, that is why I think the other approach is cleaner. But that is actually not the point I tried to make.
I think there are valid things in both approaches ; thus here is what I would suggest, so that we get to a better solution all around:
- add the page class to the PagedScrollViewController swift file
- apply the change I suggested with a static instanceFromNib() function in the Page class - this is a commonly used pattern
- use a loop in order to create the different pages and add them to the scrollview
- as you suggested, go for using frames instead of a more complex to read (and buggy, in that case, because contentSize seems to be wrong) notation.

Also, you are right that, if we want to be really clean, we will need to take care of the orientation change - layoutSubviews should indeed help there.

Reply:
Ok I redid it with frames